firedrake/assemble.py

```diff
@@ -241,6 +242,22 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
             for bc in bcs:
                 integral_types += [integral.integral_type() for integral in bc.integrals()]
 
+    new_coefficients = list(f.coefficients())
+    for ki in kernels:
+        if hasattr(ki.kinfo, 'external_operators'):
+            for k, v in ki.kinfo.external_operators.items():
+                c = f.coefficients()[k]
+                d = new_coefficients[k]
+                # Check if we need to reconstruct new ExtOps
+                if d == c._extop_master and not (len(v) == 1 and v[0] == (0,)*len(v[0])):
+                    # ExternalOperators can generate other Extops (e.g when differentiating)
+                    # The kernel contains information about which Extop has been created during form compiling,
+                    # we still need to construct this dependency when needed, i.e. when the form is already
+                    # compiled and therefore the differentiation bit of the code is not hit.
+                    new_coefficients.remove(c)
+                    d._add_dependencies(v)
+                    new_coefficients.extend([e for e in d._extop_dependencies if e.derivatives in v])
```

Comment:
So it seems like everywhere you have to special-case that this "coefficient" is not a coefficient, but rather an expression that depends on a bunch of coefficients. Why not just say "I am an expression that depends on some coefficients"?

Reply:
Or is the problem that inside the kernel there really is just one coefficient, but it must be computed in some funny way?

Reply:
Yes exactly, it is just one coefficient which requires additional procedures to deal with it, it is a coefficient enriched with additional methods.

For example, it is a coefficient that can generate other coefficients: `grad(p(u)))` where `p` is a pointwise operator will produce when hitting the form compiler: `p'(u)*grad(u)`, the form now contains `p'` and not `p`.